File: ruby/kth-smallest-element-in-a-bst.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} k
# @return {Integer}
def kth_smallest(root, k)
  index = 0
  left_root_right = lambda do |node|
    return if node.nil?

    l = left_root_right.(node.left)
    return l if l

    index += 1
    return node if index >= k

    left_root_right.(node.right)
  end

  left_root_right.(root).val
end

File: python3/kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        def left_root_right(node):
            if node is None:
                return
            yield from left_root_right(node.left)
            yield node
            yield from left_root_right(node.right)

        for i, node in enumerate(left_root_right(root), start=1):
            if i == k:
                return node.val
        return -1
